src/libfreenect2.cpp

```diff
@@ -1115,4 +1119,239 @@ Freenect2Device *Freenect2::openDefaultDevice(const PacketPipeline *pipeline)
+  packet.jpeg_buffer = frame->data;
+  packet.jpeg_buffer_length = frame->bytes_per_pixel;
+  packet.exposure = frame->exposure;
+  packet.gain = frame->gain;
+  packet.gamma = frame->gamma;
+
+  pipeline_->getRgbPacketProcessor()->process(packet);
+}
+
+void Freenect2ReplayDevice::processDepthFrame(Frame* frame)
+{
+  DepthPacket packet;
+  packet.timestamp = frame->timestamp;
+  packet.sequence = frame->sequence;
+  packet.buffer = frame->data;
+  packet.buffer_length = frame->bytes_per_pixel;
+ 
+  pipeline_->getDepthPacketProcessor()->process(packet);
+}
+
+void Freenect2ReplayDevice::loadP0Tables(unsigned char* buffer, size_t buffer_length)
+{
+  pipeline_->getDepthPacketProcessor()->loadP0TablesFromCommandResponse(buffer, buffer_length);
+}
+
+void Freenect2ReplayDevice::loadXZTables(const float *xtable, const float *ztable)
+{
+  pipeline_->getDepthPacketProcessor()->loadXZTables(xtable, ztable);
+}
+
+void Freenect2ReplayDevice::loadLookupTable(const short *lut)
+{
+  pipeline_->getDepthPacketProcessor()->loadLookupTable(lut);
+}
+
+void runner (void* arg)
+{
+  static_cast<Freenect2ReplayDevice*>(arg)->run();
+}
+
+bool Freenect2ReplayDevice::start()
+{
+  running_ = true;
+  startStreams(true, true);
+  t_ = new thread(runner, this);
+  return running_;
+}
+
+bool Freenect2ReplayDevice::startStreams(bool enable_rgb, bool enable_depth)
+{
+  LOG_INFO << "starting...: rgb: " << enable_rgb << ", depth: " << enable_depth;
+
+  if (enable_rgb)
+  {
+    LOG_INFO << "starting rgb replay...";
+  }
+
+  if (enable_depth)
+  {
+    LOG_INFO << "submitting depth replay...";
+  }
+
+  LOG_INFO << "started";
+  return true;
+}
+
+bool Freenect2ReplayDevice::stop()
+{
+  running_ = false;
+  t_->join();
+  delete t_;
+  t_ = NULL;
+  return !running_;
+}
+
+bool hasSuffix(const std::string& str, const std::string& suffix)
+{
+  if (str.length() < suffix.length())
+  {
+    return false;
+  }
+  return str.compare(str.length() - suffix.length(), suffix.length(), suffix) == 0;
+}
+
+bool Freenect2ReplayDevice::run()
+{
+  DIR *d;
+  struct dirent *dir;
+
+  std::vector<std::string> frames;
+
+  d = opendir(directory_);
+  while ((dir = readdir(d)) != NULL)
+  {
+    std::string name = dir->d_name;
+    if (hasSuffix(name, ".depth")) {
+      frames.push_back(name);
+    }
+  }
+
+  for (size_t i = 0; i < frames.size() && running_; i++) {
+    std::string frame = frames[i];
+
+    size_t ix1 = frame.find("_");
+    size_t ix2 = frame.find("_", ix1 + 1);
+    size_t ix3 = frame.find(".", ix2 + 1);
+
+    std::string ts = frame.substr(0, ix1);
+    std::string seq = frame.substr(ix2 + 1, ix3);
+
+    std::string data;
+    if (hasSuffix(frame, ".depth")) {
+      std::ifstream fd(frame.c_str());
+      fd.seekg(0, fd.end);
+      int length = fd.tellg();
+      fd.seekg(0, fd.beg);
+
+      DepthPacket packet;
+      packet.timestamp = atoi(ts.c_str());
+      packet.sequence = atoi(seq.c_str());
+      packet.buffer = new unsigned char[length];
```

Comment:
Memory allocation is potentially expensive. Do you have to allocate new memory here? My idea would be
```cpp
std::vector<char> buffer;
buffer.reserve(size_of_a_depth_frame);
...
buffer.resize(length);
packet.buffer = reinterpret_cast<unsigned char*>(&buffer[0]);
fd.read(&buffer[0], length);
```

Reply:
Simply inherited the original from #555, I am OK with your idea; either way memory is allocated somewhere via new or resize...

Reply:
`resize()` does not allocate if `size` is not larger than `capacity`. This way it just allocates once for all packets.

Reply:
At that time in `run()` I do not know `size_of_a_depth_frame` except that the file length determines it. So the actual code looks like:
```cpp
      packet.buffer_length = length;
      std::vector<char> buffer;
      buffer.reserve(length);
      packet.buffer = reinterpret_cast<unsigned char*>(&buffer[0]);
      fd.read(&buffer[0], length);
```

Reply:
The size of a raw frame is here:

https://github.com/OpenKinect/libfreenect2/blob/master/src/depth_packet_stream_parser.cpp#L43

Actually, I think you need to take a look at DepthPacketStreamParser in general: you should allocate the fixed size buffer at initialization, and processor should be protected by calling ready(), and allocate again after process() (there is internal double buffering). You allocate by calling `processor_->allocateBuffer(...)` which calls the internal memory pool allocator.

Reply:
TODO